### PR TITLE
Fix truncate behavior

### LINKED ISSUE #9: Update delete logic to be soft-delete
Check guide.
Keep diff around because hard deletes will eventually be supported.

From the guide:
> Truncate events after update events are normal. This is most likely an initial sync where there is nothing but upsert operations, all followed by a truncate which is meant to (soft) delete any rows that may have existed prior to the initial sync starting to make sure they are all marked as deleted (since we cannot be sure the initial sync will necessarily overwrite them all). The "before timestamp" is key to the truncate operation so you don't just mark the entire table deleted. It should pick out the rows that existed prior to the sync starting (aka _fivetran_synced < "truncate before timestamp")

### LINKED ISSUE #15: Truncate has to take utc_delete_before into account
Right now, truncate is a blanket truncate. It should instead only delete the records prior to the `utc_delete_before` date passed in.

https://github.com/fivetran/fivetran_sdk/blob/main/development-guide.md#is-it-normal-that-for-a-sync-there-is-an-upsert-event-followed-by-a-truncate-event

### PR DESCRIPTION
**Pull request by elefeint**
Two fixes for truncate behavior:
- hard deletes are not supported yet; all truncates have to be soft-delete
- truncate should take into account the column / value for `utc_delete_before` (timestamp prior to which to delete rows).

Fixes #15 
Fixes #9

### CODE REVIEW

File: src/motherduck_destination_server.cpp [hunk 1]
```diff
@@ -269,11 +269,22 @@ DestinationSdkImpl::Truncate(::grpc::ServerContext *context,
...
+      // right now, only soft deletes are supported
+      throw std::invalid_argument("Deleted column is required");
+    }
     std::unique_ptr<duckdb::Connection> con =
         get_connection(request->configuration(), db_name);
 
     if (table_exists(*con, table_name)) {
-      truncate_table(*con, table_name);
+      std::chrono::nanoseconds delete_before_ts =
+          std::chrono::seconds(request->utc_delete_before().seconds()) +
```
[the remark this answers is not in the document]
Comment (elefeint) [reply src/motherduck_destination_server.cpp:284]: Adding seconds to nanoseconds gets seconds converted automatically ([docs](https://cplusplus.com/reference/chrono/duration/operators/))
> When two [duration](https://cplusplus.com/duration) objects of different types are involved, the one with the longest period (as determined by [common_type](https://cplusplus.com/common_type)) is converted before the operation.

`nanos()` only contain the fractional part -- [proto docs.](https://github.com/protocolbuffers/protobuf/blob/7949a62bf88e1d4e97c93b3a7b8a8cfb201b713a/src/google/protobuf/timestamp.proto#L90)

File: src/sql_generator.cpp [hunk 2]
```diff
@@ -319,13 +319,27 @@ void delete_rows(duckdb::Connection &con, const table_def &table,
...
-void truncate_table(duckdb::Connection &con, const table_def &table) {
+void truncate_table(duckdb::Connection &con, const table_def &table,
+                    const std::string &synced_column,
+                    std::chrono::nanoseconds &cutoff_ns,
+                    const std::string &deleted_column) {
   const std::string absolute_table_name = table.to_string();
   std::ostringstream sql;
-  sql << "DELETE FROM " + absolute_table_name;
+
+  sql << "UPDATE " << absolute_table_name << " SET "
```
[the remark this answers is not in the document]
Comment (elefeint) [reply src/sql_generator.cpp:329]: `table_def::to_string()` quotes all parts

[the remark this answers is not in the document]
Comment (elefeint) [reply src/sql_generator.cpp:329]: perhaps I should rename it to `to_quoted_string`

[the remark this answers is not in the document]
Comment (elefeint) [reply src/sql_generator.cpp:330]: renamed to `to_escaped_string`

**Review by Y-- — APPROVED** [review]
LGTM!

File: src/motherduck_destination_server.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (elefeint) [reply src/motherduck_destination_server.cpp:284]: I was actually grateful that for once C++ did not make me invent every basic utility from first principles, but yeah, it's not super intuitive either.
And I have no idea what proto designers were thinking here. I guess it's space saving for usecases where only second granularity is needed? But who uses second granularity anymore -- it's not 1999.

**Pull request merged**